crates/ruma-common/src/canonical_json.rs

```diff
@@ -0,0 +1,367 @@
+//! Canonical JSON types and related functions.
+
+use std::{fmt, mem};
+
+use serde::Serialize;
+use serde_json::Value as JsonValue;
+
+mod value;
+
+use crate::RoomVersionId;
+
+pub use self::value::{CanonicalJsonObject, CanonicalJsonValue};
+
+/// The set of possible errors when serializing to canonical JSON.
+#[cfg(feature = "canonical-json")]
+#[derive(Debug)]
+#[allow(clippy::exhaustive_enums)]
+pub enum CanonicalJsonError {
+    /// The numeric value failed conversion to js_int::Int.
+    IntConvert,
```

Comment:
One vague concern/idea i had when looking at this is that there's 0 tracability of which exact field failed int conversion when this happens.

This should probably be a separate issue (as this PR just moves the code around), but i'm noting it here.

---

I see `field` in `NotOfType` below, so maybe an approach like that could be copied?